### PR TITLE
Companion Linux: build AppImage

### PR DESCRIPTION
**Pull request by raphaelcoeffic**

### CODE REVIEW

**Comment by pfeerick** [comment]
Well, I think that's that issue put to rest! 😁 Was in a VM for now, but had access to real transmitter, and was able to both read from and write to it. Just need the simulators now! :)

Screenshot_20210929_152843

**Comment by raphaelcoeffic** [comment]
> Well, I think that's that issue put to rest! 😁 Was in a VM for now, but had access to real transmitter, and was able to both read from and write to it. Just need the simulators now! :)
> 
> Screenshot_20210929_152843

Dfu-util isn’t missing? I’d bet the dependencies are not packaged…

**Comment by pfeerick** [comment]
You'd win that bet... it's looking for the system installed dfu-util... 
image

**Comment by raphaelcoeffic** [comment]
@pfeerick no clue which consequences it has, but there is no way to copy the udev rules from within an AppImage... (see https://github.com/AppImage/pkg2appimage/issues/200)

**Comment by pfeerick** [comment]
I wouldn't worry about that so much in this first instance - and I'm not surprised, to do anything `udev` related you'd need root/admin privileges since you're messing around with system rules. In this case, the rules in `/companion/targets/linux` are simply setting permissions, so that the owner-group-users all have read and write access, meaning Companion doesn't need to be run as a admin user to access the taranis or usbasp.

Worst case scenario, we come up with a shell script the user runs on their system, which is able to detect the difference between debian/arch/suse systems, run the appropriate package manager commands to install `dfu-utils` for that distro, and apply any udev rules necessary. The likelihood is you won't have to hand-hold Linux users that much (else they shouldn't be using Linux in the first place!) - just tell them they need to install `dfu-util` (if they can't figure it out from the error message), and

**Comment by raphaelcoeffic** [comment]
> I wouldn't worry about that so much in this first instance - and I'm not surprised, to do anything `udev` related you'd need root/admin privileges since you're messing around with system rules. In this case, the rules in `/companion/targets/linux` are simply setting permissions, so that the owner-group-users all have read and write access, meaning Companion doesn't need to be run as a admin user to access the taranis or usbasp.
> 
> Worst case scenario, we come up with a shell script the user runs on their system, which is able to detect the difference between debian/arch/suse systems, run the appropriate package manager commands to install `dfu-utils` for that distro, and apply any udev rules necessary. The likelihood is you won't have to hand-hold Linux users that much (else they shouldn't be using Linux in the first place!) - just tell them they need to install `dfu-util` (if they can't figure it out from the error message), and

`dfu-util` needs to be installed in the AppImage, otherwise it cannot be executed from Companion. Imagine a AppImage as a sort of a sandbox ;-)

**Comment by pfeerick** [comment]
No, not quite... AppImages aren't sandboxes (like a virtual machine)... it's a self-contained packaged that can *also* access host resources and files... i.e. I just did an `apt install dfu-util` on the host OS (Ubuntu), and now instead of an error, I can flash the firmware...

image

File: companion/src/simulation/simulatorinterface.cpp [hunk 1]
```diff
@@ -100,7 +100,11 @@ void SimulatorLoader::registerSimulators()
...
   dir.setPath(path);
 #else
-  dir.setPath(SIMULATOR_LIB_SEARCH_PATH);
+  if (QDir::isAbsolutePath(SIMULATOR_LIB_SEARCH_PATH)) {
+    dir.setPath(SIMULATOR_LIB_SEARCH_PATH);
+  } else {
+    dir.setPath(QCoreApplication::applicationDirPath() + "/" SIMULATOR_LIB_SEARCH_PATH);
+  }
 #endif
   registerSimulators(dir);
```
Comment (pfeerick) [inline companion/src/simulation/simulatorinterface.cpp:109]: Would this not be better as `registerSimulators(dir.absolutePath());`?

i.e this is the current output
```
[D] SimulatorLoader::registerSimulators():55 - [simulator.interface.loader] Searching for simulators in "/tmp/.mount_Compan6V9ZnK/local/bin" matching pattern ("*-simulator.so")
[D] SimulatorLoader::registerSimulators():83 - [simulator.interface.loader] Found libraries: ("none")
[D] SimulatorLoader::registerSimulators():55 - [simulator.interface.loader] Searching for simulators in "/tmp/.mount_Compan6V9ZnK/local/bin/../lib/companion26" matching pattern ("*-simulator.so")
[D] SimulatorLoader::registerSimulators():60 - [simulator.interface.loader] Trying to register simulator in  "libopentx-nv14-simulator.so"
```
Whereas absolutepath ensures it becomes an, well, absolute path... so it's a clean path now...
```
[D] SimulatorLoader::registerSimulators():55 - [simulator.interface.loader] Searching for simulators in "/tmp/.mount_CompanQ55QXC/local/bin" matching pattern ("*-simulator.so")
[D] SimulatorLoader::registerSimulators():83 - [simulator.interface.loader] Found libraries: ("none")
[D] SimulatorLoader::registerSimulators():55 - [simulator.interface.loader] Searching for simulators in "/tmp/.mount_CompanQ55QXC/local/lib/companion26" matching pattern ("*-simulator.so")
[D] SimulatorLoader::registerSimulators():60 - [simulator.interface.loader] Trying to register simulator in  "libopentx-tx16s-simulator.so"
```

**Comment by pfeerick** [comment]
Regardless, this package is working fine for me on Windows and Linux (Ubuntu), and seems to be able to read and write to the radio properly in both cases. Simulators also both run for both.

File: companion/src/simulation/simulatorinterface.cpp (hunk 1 above)
Comment (raphaelcoeffic) [reply companion/src/simulation/simulatorinterface.cpp]: I changed it ;-) Let's see how that works.

**Comment by raphaelcoeffic** [comment]
@pfeerick by the way, I assume that the code for windows is basically trying to do the same thing with some strange Win32 API:
```c++
  char name[MAX_PATH];
  GetModuleFileName(NULL, name, MAX_PATH);
  QString path(name);
  path.truncate(path.lastIndexOf('\\'));
  dir.setPath(path);
```

I think we should try to get rid of this at the same time. Is that something you could try out?

**Review by pfeerick — APPROVED** [review]
I think Risto has identified the dfu-utils problems to be a combination of host-install path issues and a underlying bug in companion which has been since fixed. 

I haven't had any issues with this when trying it on Linux - either in a VM or on actual hardware.

**Pull request merged**